src/extra/components/adq.py

```diff
@@ -0,0 +1,1318 @@
+                                     'length': samples_per_pulse})
+
+        return pulse_ids, pulse_layout
+
+    def _reshape_pulses_by_train(self, data, num_pulses, samples_per_pulse,
+                                 first_pulse_offset=None):
+        """Reshape traces by train into traces by train and pulses."""
+
+        pulses_start = first_pulse_offset or self._first_pulse_offset
+        pulses_end = pulses_start + num_pulses * samples_per_pulse
+
+        if pulses_end > data.shape[-1]:
+            raise ValueError(f'trace axis too short for {num_pulses} pulses '
+                             f'located at {pulses_start}:{pulses_end}]')
+
+        return data[..., pulses_start:pulses_end].reshape(
+            *data.shape[:-1], num_pulses, samples_per_pulse)
+
+    def _reshape_flat_pulses(self, data, out, pulse_ids, samples_per_pulse,
+                             first_pulse_offset=None):
+        """Reshape traces by train into traces by pulses."""
+
+        return _reshape_flat_pulses(
+            data, out, pulse_ids, samples_per_pulse,
+            first_pulse_offset or self._first_pulse_offset, self._clock_ratio)
+
+    def _build_sample_coords(self, data):
+        """Build xarray coordinates for samples."""
+
+        samples = np.arange(data.shape[-1], dtype=np.int32)
+
+        if self._sample_dim == 'time':
+            return {'time': samples * self.sampling_period}
+        elif self._sample_dim == 'sample':
+            return {'sample': samples}
+
+    def _shape_edges(self, edges, amplitudes, index):
+        """Build pandas object from edge/amplitude array."""
+
+        from . import DelayLineDetector
+
+        edges = DelayLineDetector._build_reduced_pd(
+            None, edges, index, entry_level='edgeIndex')
+        amplitudes = DelayLineDetector._build_reduced_pd(
+            None, amplitudes, index, entry_level='edgeIndex')
+
+        return pd.DataFrame({'edge': edges, 'amplitude': amplitudes})
+
+    def _shape_edge_array(self, edges, amplitudes, coords,
+                          labelled, squeeze_edges):
+        """Shape output data of edge array finding methods."""
+
+        if squeeze_edges:
+            edges = self._minimize_sparse_array(edges)
+            amplitudes = self._minimize_sparse_array(amplitudes)
+
+        if not labelled:
+            return edges, amplitudes
+
+        if coords is None:
+            # Generate generic coordinates.
+            coords = {f'dim_{i}': np.arange(edges.shape[i])
+                      for i in range(edges.ndim - 1)}
+
+        coords['edge'] = np.arange(edges.shape[-1])
+
+        import xarray as xr
+        return xr.Dataset(dict(
+            edges=xr.DataArray(edges, coords=coords),
+            amplitudes=xr.DataArray(amplitudes, coords=coords)))
+
+    def _iter_cont_chunks(self):
+        """Iterate over contiguous dataset regions."""
+
+        out_start = 0
+
+        # Calling private EXtra-data APIs, use with caution.
```

Comment:
Can I again ask that we either switch this to using `.split_trains()`  to make chunks or just do the correction on all the loaded data at once for now?

'Use with caution' is kind of a meaningless comment. The risk with using private APIs is that they change without warning and break the code using them. Caution in writing the downstream code can't prevent that.

Reply:
It's there for us to realize it's a private API and should not be used lightly. While it's always better to rest on public APIs, I do want to be able to use private APIs in EXtra-data. If we change it, we have to change it in EXtra as well. We should always have tests to tell us that.

I left this part in to save time. I fully agree this is not robust for virtual overview files and needs to change, but they are not used at SQS, and there is little operational pressure to change implementation details.

Reply:
It's fine, I meant to do some profiling first to find an equally efficient solution, but any likely performance loss is acceptable now. It looks like I can even maintain the cast-on-copy behaviour, which was the original reason I went this route.

Reply:
Please have a look. For now I simply hardcoded `trains_per_part=200`, which for average trace lengths should yield ~500 MiB data per iteration. This could be done smarter in the future.

Reply:
Thanks! I'm happy with that.